### ISSUE TITLE
CDC: within a batch, partition deletes and range deletes do not affect postimage

### ISSUE THREAD

**Issue by piodul** [scylladb/scylla#6900]
A CQL batch that both

1. inserts or updates a row `X` with timestamp `T`,
2. deletes that row using row delete or partition delete that covers the row `X`, also using timestamp `T`,

has wrong postimage for the row `X`. Current logic shows the value set by 1) and ignores the effect of 2).

It is wrong, because range/partition delete wins with the insert/update, and cancels its effect - therefore postimage in this case does not reflect the actual state of the row after the batch is applied.

Example with insert + range delete:

```cql
cqlsh:ks> create table tbl (pk int, ck int, v int, primary key(pk, ck)) with cdc = {'enabled': true, 'preimage': true, 'postimage': true};
cqlsh:ks> begin batch insert into tbl (pk, ck, v) values (1, 1, 100); delete from tbl where pk = 1 and ck >= 1 and ck <= 2; apply batch;
cqlsh:ks> select * from tbl;

 pk | ck | v
----+----+---

(0 rows)
cqlsh:ks> select * from tbl_scylla_cdc_log;

 cdc$stream_id                      | cdc$time                             | cdc$batch_seq_no | cdc$deleted_v | cdc$operation | cdc$ttl | ck | pk | v
------------------------------------+--------------------------------------+------------------+---------------+---------------+---------+----+----+------
 0xc74d27bc17a40015893bffbb671eb09d | 6743d17a-cc0f-11ea-8c1b-e371909c58d6 |                0 |          null |             0 |    null |  1 |  1 | null
 0xc74d27bc17a40015893bffbb671eb09d | 6743d17a-cc0f-11ea-8c1b-e371909c58d6 |                1 |          null |             2 |    null |  1 |  1 |  100
 0xc74d27bc17a40015893bffbb671eb09d | 6743d17a-cc0f-11ea-8c1b-e371909c58d6 |                2 |          null |             5 |    null |  1 |  1 | null
 0xc74d27bc17a40015893bffbb671eb09d | 6743d17a-cc0f-11ea-8c1b-e371909c58d6 |                3 |          null |             7 |    null |  2 |  1 | null
 0xc74d27bc17a40015893bffbb671eb09d | 6743d17a-cc0f-11ea-8c1b-e371909c58d6 |                4 |          null |             9 |    null |  1 |  1 |  100

(5 rows)
```

Example with insert + partition delete:

```cql
cqlsh:ks> create table tbl (pk int, ck int, v int, primary key(pk, ck)) with cdc = {'enabled': true, 'preimage': true, 'postimage': true};
cqlsh:ks> begin batch insert into tbl (pk, ck, v) values (1, 1, 100); delete from tbl where pk = 1; apply batch;
cqlsh:ks> select * from tbl;

 pk | ck | v
----+----+---

(0 rows)
cqlsh:ks> select * from tbl_scylla_cdc_log;

 cdc$stream_id                      | cdc$time                             | cdc$batch_seq_no | cdc$deleted_v | cdc$operation | cdc$ttl | ck   | pk | v
------------------------------------+--------------------------------------+------------------+---------------+---------------+---------+------+----+------
 0xc74d27bc17a40015893bffbb671eb09d | bd56e1f6-cc0f-11ea-67fc-ead359feab5f |                0 |          null |             0 |    null |    1 |  1 | null
 0xc74d27bc17a40015893bffbb671eb09d | bd56e1f6-cc0f-11ea-67fc-ead359feab5f |                1 |          null |             2 |    null |    1 |  1 |  100
 0xc74d27bc17a40015893bffbb671eb09d | bd56e1f6-cc0f-11ea-67fc-ead359feab5f |                2 |          null |             4 |    null | null |  1 | null
 0xc74d27bc17a40015893bffbb671eb09d | bd56e1f6-cc0f-11ea-67fc-ead359feab5f |                3 |          null |             9 |    null |    1 |  1 |  100

(4 rows)
```

In both cases, it can be seen that the inserted row does NOT appear in base table, but the postimage suggests otherwise (it should be `null`, but is `100`).

**Comment by haaawk** [comment]
@elcallio This is the issue we've been talking about on the status call

**Comment by nyh** [comment]
Not backporting to 4.2. CDC is still experimental, and the code involved in this fix was refactoredy since 4.2 so backporting it would require to backport the refactor, which I don't want to do.
